### PR TITLE
[FEATURE] Correct and improve vcf output (insertion calling 2/2)

### LINKED ISSUE #111: [FEATURE,FIX] Correct and improve VCF output
- [x] The VCF header line `CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO` is missing the preceding `#`.
- [x] Output INS calls based on the junctions (in `src/variant_detection/variant_output`)
- [x] Minimum variant size is currently hard-coded in `find_and_output_variants()`. We should use the command-line argument instead.
- [x] The coordinates returned by [record.reference_position](https://docs.seqan.de/seqan/3-master-dev/classseqan3_1_1sam__record.html#a959338eb203f0360499056f1530d7f4f) are 0-based but the POS field in VCF needs to be 1-based. Therefore, we need to add 1 to every coordinate in `find_and_output_variants()`.
- [x] Deletion junctions specify the last base `a` before and the first base `b` after the deletion. To output correct coordinates in VCF, we need to set `POS=a` and `END=b-1`, I believe.
- [x] Insertion junctions specify the last base `a` before and the first base `b=a+1` after the insertion. To output correct coordinates in VCF, we need to set `POS=a` and `END=a`, I believe.
- [x] The QUAL field of the output VCF is hard-coded to 60 for all calls. Instead, we can use the cluster size as a preliminary score for the calls.

### PR DESCRIPTION
**Pull request by eldariont**
This PR fixes #111 and adds the output of insertions into the VCF. Additionally, it correct several small errors in the VCF output and propagates the `min_var_length` parameter to the VCF output method.

### CODE REVIEW

**Review by MitraDarja — APPROVED** [review]

**Review by Irallia — CHANGES_REQUESTED** [review]
I'm happy to see, that we got an output file now! :tada:
Thanks for the corrections and their documentation in the code.
I have some comments for the `find_and_output_variants()` and also thougt, that we should start testing edge cases.

File: include/variant_detection/variant_output.hpp [hunk 1]
```diff
@@ -4,26 +4,34 @@
...
+ * \param[in]       args        - command line arguments:\n
+ *                                **args.min_var_length** - minimum length of variants to detect - *default: 30 bp*
  * \param[in, out]  out_stream  - output stream
  *
  * \details Extracts genomic variants from given junction clusters.
  */
-void find_and_output_variants(std::vector<Cluster> const & clusters, std::ostream & out_stream);
+void find_and_output_variants(std::vector<Cluster> const & clusters,
+                              cmd_arguments const & args,
+                              std::ostream & out_stream);
```
[the remark this answers is not in the document]
Comment (eldariont) [reply include/variant_detection/variant_output.hpp:22]: Yeah, I thought about that for a while, too, and ended up passing the args because this method handles the actual variant calling and output. In the future, we might add a lot of arguments that need to be passed to the method, e.g.
- maximum variant length
- maximum tolerated inserted sequence size for deletions (currently hard-coded as 5)
- maximum tolerated distance between breakends for insertions (current hard-coded as 1)
- variant classes to output
- output of symbolic or sequences alleles
- ...

File: src/iGenVar.cpp [hunk 2]
```diff
@@ -9,17 +9,6 @@
...
-{
-    std::filesystem::path alignment_short_reads_file_path{""};
-    std::filesystem::path alignment_long_reads_file_path{""};
-    std::filesystem::path output_file_path{};
-    std::vector<detection_methods> methods{cigar_string, split_read, read_pairs, read_depth};   // default: all methods
-    clustering_methods clustering_method{simple_clustering};                                    // default: simple clustering method
-    refinement_methods refinement_method{no_refinement};                                        // default: no refinement
-    uint64_t min_var_length = 30;
-};
-
```
[the remark this answers is not in the document]
Comment (eldariont) [reply src/iGenVar.cpp:22]: Regardless of whether we pass args to `find_and_output_variants()` or not: I wondered why the struct was defined in the cpp and not the hpp. Would it not be more common to define the fields of a struct in the hpp?

File: src/variant_detection/variant_output.cpp [hunk 3]
```diff
@@ -16,41 +19,49 @@ void find_and_output_variants(std::vector<Cluster> const & clusters, std::ostrea
...
                 int32_t mate2_pos = mate2.position;
+                int32_t insert_size = clusters[i].get_average_inserted_sequence_size();
                 if (mate1.orientation == strand::forward)
                 {
                     int32_t distance = mate2_pos - mate1_pos;
-                    if (distance > 40 && distance < 100000)
+                    //Deletion
+                    if (distance >= args.min_var_length &&
+                        distance < 1000000 &&
+                        insert_size < 5)
```
[the remark this answers is not in the document]
Comment (eldariont) [reply src/variant_detection/variant_output.cpp:36]: Good point, we should get rid of this hard-coded value and replace it with a CLI parameter. Yes, it means that we accept up to 4 insertions in a deletion, e.g.
```
ABCDEFGHI
AB-----HI  deletion of five bases
  RST      insertion of three bases instead
```

File: src/variant_detection/variant_output.cpp [hunk 4]
```diff
@@ -16,41 +19,49 @@ void find_and_output_variants(std::vector<Cluster> const & clusters, std::ostrea
...
+                        tmp.set_qual(cluster_size);
                         tmp.set_alt("<DEL>");
                         tmp.add_info("SVTYPE", "DEL");
-                        tmp.set_pos(mate1_pos);
-                        tmp.add_info("SVLEN", std::to_string(-distance));
+                        // Increment position by 1 because VCF is 1-based
+                        tmp.set_pos(mate1_pos + 1);
+                        tmp.add_info("SVLEN", std::to_string(-distance + 1));
+                        // Increment end by 1 because VCF is 1-based
+                        // Decrement end by 1 because deletion ends one base before mate2 begins
```
[the remark this answers is not in the document]
Comment (eldariont) [reply src/variant_detection/variant_output.cpp:47]: I totally agree 👍

File: src/variant_detection/variant_output.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (eldariont) [reply src/variant_detection/variant_output.cpp:36]: Added this in separate commit.

File: src/variant_detection/variant_output.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (eldariont) [reply src/variant_detection/variant_output.cpp:47]: Created an issue: #117

**Comment by eldariont** [comment]
Hi @Irallia, I rebased this PR after the recent merges and incorporated your last three suggestions. Please let me know what you think about the other two suggestions where I added my reply 🙂

File: include/variant_detection/variant_output.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Irallia) [reply include/variant_detection/variant_output.hpp:25]: okay, thats a good argument. Thanks for adding some of these variables directly :)

File: src/iGenVar.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Irallia) [reply src/iGenVar.cpp:22]: Yes you are right. The other structs we have also in the header files.

**Review by Irallia — APPROVED** [review]
